### PR TITLE
FEAT: Add Azure Function (Python) sample (correlated E2E monitoring)

### PR DESCRIPTION
**Pull request by se02035**
# What we are trying to address

Providing a sample implementation demonstrating how to instrument Azure Function (Python) code using Opencensus to get a full end2end view of the function invocation in Azure Monitor (Application Insights). 

Considering that the out-of-the-box (OOTB) logging/Opencensus integration in Azure Function has some open issues, this sample helps developers to create functions with a better level of observability. Also developers have a small but complete sample that will support them in better understanding the integration with Opencensus.

## Description of Changes

- Added a new monitoring sample under *"azure_monitor/azfunc_sample"*. it contains
    - Azure function
    - Tests
    - Documentation

## How did we test it

- Unit/Integration tests 
- Sending sample requests using curl and verify output in configured Application Insights instance

## Any relevant logs or outputs

AppInsights E2E function invocation overview (correct correlation)
image

AppInsights Application Map overview (correct role & instance name)
image

### CODE REVIEW

File: azure_monitor/azfunc_sample/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+- azure
+description: "This sample contains a simple Azure Function to show how you can send correlated application log information as well as information of external dependencies to Azure Monitor (Application Insights)."
+urlFragment: azure-monitor-opencensus-python
+---
+
+# Correlated End-To-End Azure Function (Python) Instrumentation with Opencensus 
+
+## Overview
+
+Azure Function (Python) already supports instrumentaton using Opencensus. Unfortunately, the current out-of-the-box (OOTB) implementation/support has some flaws (see below), which make it difficult to get a correlated end-to-end view of a single function invocation (in Azure Monitor/Application Insights).
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:15]: I would not say "flaws" nor "unfortunately", as this refers to a bug of some kind. Maybe reword this to: `The following sample utilizes the OpenCensus Functions extension and enables additional common telemetry scenarios used by Azure Monitor and Functions users. Such scenarios include:...`

File: azure_monitor/azfunc_sample/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+![AI Application trace messages](./docs/assets/02-ai-correlated-invocation.PNG)
+
+Associated AI Application Map:
+
+![AI Application Map](./docs/assets/03-ai-application-map.PNG)
+
+The images shows the following elements:
+
+- Correlated (per Azure Function invocatio) log information incl
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:35]:
```suggestion
- Correlated (per Azure Function invocation) log information including:
```

File: azure_monitor/azfunc_sample/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    - Incoming request information
+    - Appication traces & error messages
+    - External dependencies (IN-PROC & HTTP using Python's requests)
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:38]:
```suggestion
    - External dependencies (IN-PROC & HTTP using Python's `requests` library)
```

File: azure_monitor/azfunc_sample/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+Associated AI Application Map:
+
+![AI Application Map](./docs/assets/03-ai-application-map.PNG)
+
+The images shows the following elements:
+
+- Correlated (per Azure Function invocatio) log information incl
+    - Incoming request information
+    - Appication traces & error messages
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:37]:
```suggestion
    - Application traces & error messages
```

File: azure_monitor/azfunc_sample/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    - External dependencies (IN-PROC & HTTP using Python's requests)
+- Application Map (defined cloud role and cloud instance values)
+
+### Known Limitations in Current OOTB Logging Integration (Azure Function)
+
+- External dependency log records don't have the correct *Cloud Role Name* (field: cloud_RoleName). This results in wrong information displayed in Application Insight's *Application Map*. See <https://github.com/census-instrumentation/opencensus-python/issues/1052>
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:43]: This is not a limitation. Users can configure cloud role name via [telemetry processors](https://github.com/census-instrumentation/opencensus-python/issues/1052)

File: azure_monitor/azfunc_sample/README.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+The images shows the following elements:
+
+- Correlated (per Azure Function invocatio) log information incl
+    - Incoming request information
+    - Appication traces & error messages
+    - External dependencies (IN-PROC & HTTP using Python's requests)
+- Application Map (defined cloud role and cloud instance values)
+
+### Known Limitations in Current OOTB Logging Integration (Azure Function)
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:41]:
```suggestion
### Known Limitations in Current OOTB Logging Integration (OpenCensus Azure Monitor log exporter)
```

File: azure_monitor/azfunc_sample/README.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+- External dependency log records don't have the correct *Cloud Role Name* (field: cloud_RoleName). This results in wrong information displayed in Application Insight's *Application Map*. See <https://github.com/census-instrumentation/opencensus-python/issues/1052>
+- Errors (logged with the OOTB logging integration) are not correctly logged as Application Insights Exception records. Currently `logging.exception(..)` creates a trace record (with severity level 3 = Error). This comes with certain side effects: E.g. errors are won't show up in Application Insights dashboard (*Failures*) and troubleshooting becomes harder (see issue <https://github.com/Azure/azure-functions-python-worker/issues/866>)
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:44]: If we are going to use `OOTB logging integration` please include an excerpt at the beginning that explicit says that it is referring to the [OpenCensus Azure monitor log exporter](https://github.com/census-instrumentation/opencensus-python/blob/master/contrib/opencensus-ext-azure/opencensus/ext/azure/log_exporter/__init__.py).

File: azure_monitor/azfunc_sample/README.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+## Goal 
+
+The goal of this code sample is to demonstrate an alternative (custom) approach of instrumenting Azure Function code using Opencensus. This should give developers more insights how to use Opencensus and how to overcome the limitations of the OOTB implementation.
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:48]: Like above, please frame this as "customizing" instead of "limitations".

File: azure_monitor/azfunc_sample/instrumentation/globals.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
+import os
+import socket
+import logging
+
+from logging import Logger
+from opencensus.trace import config_integration
+from opencensus.ext.azure.log_exporter import AzureLogHandler
+
+
+INSTRUMENTATION_KEY = os.getenv("APPLICATIONINSIGHTS_CONNECTION_STRING")
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/instrumentation/globals.py:10]:
```suggestion
CONNECTION_STRING = os.getenv("APPLICATIONINSIGHTS_CONNECTION_STRING")
```

File: azure_monitor/azfunc_sample/instrumentation/globals.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+if not INSTRUMENTATION_KEY:
+    raise EnvironmentError(
+        "Instrumentation Key not set. Set it through the environment variable: APPINSIGHTS_INSTRUMENTATIONKEY."
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/instrumentation/globals.py:14]:
```suggestion
        "Connection string not set. Set it through the environment variable: APPLICATIONINSIGHTS_CONNECTION_STRING."
```

File: azure_monitor/azfunc_sample/README.md (hunk 7 above)
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:44]: Seems like `OOTB logging integration` in this section is referring to the Azure functions worker logger, not the Opencensus logger. The `exceptions` limitation is an issue in the functions logger, but the OpenCensus logger does not have this issue. With that being said, users will be getting duplicate telemetry if they start using the OpenCensus logger as well (since both loggers send user generated logs). We are developing a feature to disable this double telemetry.

File: azure_monitor/azfunc_sample/README.md (hunk 1 above)
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:15]:
```suggestion
Azure Function (Python) already supports instrumentation using OpenCensus. Unfortunately, the current out-of-the-box (OOTB) implementation/support has some flaws (see below), which make it difficult to get a correlated end-to-end view of a single function invocation (in Azure Monitor/Application Insights).
```

File: azure_monitor/azfunc_sample/README.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+The images shows the following elements:
+
+- Correlated (per Azure Function invocatio) log information incl
+    - Incoming request information
+    - Appication traces & error messages
+    - External dependencies (IN-PROC & HTTP using Python's requests)
+- Application Map (defined cloud role and cloud instance values)
+
+### Known Limitations in Current OOTB Logging Integration (Azure Function)
+
```
Comment (lzchen) [inline azure_monitor/azfunc_sample/README.md:42]: Another point is that the OOTB implementation does not allow for log correlation with the distributed tracing telemetry that it is part of (which is what your `config_integration.trace_integrations(["logging"])` line does.

File: azure_monitor/azfunc_sample/README.md (hunk 7 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:44]: thanks for the update. I noticed the "duplicated log entry" issue. therefore, I explicitly disable log propagation in the sample code. but good to know that there will be a feature in future.

File: azure_monitor/azfunc_sample/README.md (hunk 2 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:35]: done. updated.

File: azure_monitor/azfunc_sample/README.md (hunk 3 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:38]: done. updated.

File: azure_monitor/azfunc_sample/README.md (hunk 4 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:37]: done. updated.

File: azure_monitor/azfunc_sample/README.md (hunk 6 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:41]: done. updated.

File: azure_monitor/azfunc_sample/instrumentation/globals.py (hunk 9 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/instrumentation/globals.py:10]: done. updated.

File: azure_monitor/azfunc_sample/instrumentation/globals.py (hunk 10 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/instrumentation/globals.py:14]: done. updated

File: azure_monitor/azfunc_sample/README.md (hunk 8 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:48]: done. updated.

File: azure_monitor/azfunc_sample/README.md (hunk 7 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:44]: replaced OOTB logging integration with "OpenCensus Azure Monitor Log Exporter"

File: azure_monitor/azfunc_sample/README.md (hunk 1 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:15]: done. updated.

Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:15]: done. updated

File: azure_monitor/azfunc_sample/README.md (hunk 11 above)
Comment (se02035) [reply azure_monitor/azfunc_sample/README.md:42]: true. will add that.

**Comment by se02035** [comment]
@lzchen , do you have any other comments related to this PR? Or do you think it can get merged.

**Comment by se02035** [comment]
@lzchen , any updates on this? do you think that can get merged?

**Review by lzchen — APPROVED** [review]
